File: src/main/java/org/taktik/icure/logic/impl/filter/contact/ContactByHcPartyTagCodeDateFilter.java
/*
 * Copyright (C) 2018 Taktik SA
 *
 * This file is part of iCureBackend.
 *
 * iCureBackend is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as published by
 * the Free Software Foundation.
 *
 * iCureBackend is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with iCureBackend.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.taktik.icure.logic.impl.filter.contact;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.taktik.icure.entities.Contact;
import org.taktik.icure.entities.User;
import org.taktik.icure.logic.ContactLogic;
import org.taktik.icure.logic.ICureSessionLogic;
import org.taktik.icure.logic.impl.filter.Filter;
import org.taktik.icure.logic.impl.filter.Filters;

import javax.security.auth.login.LoginException;
import java.util.HashSet;
import java.util.List;
import java.util.Set;


@Service
public class ContactByHcPartyTagCodeDateFilter implements Filter<String, Contact, org.taktik.icure.dto.filter.contact.ContactByHcPartyTagCodeDateFilter> {
	ContactLogic contactLogic;
	ICureSessionLogic sessionLogic;

	@Autowired
	public void setContactLogic(ContactLogic contactLogic) {
		this.contactLogic = contactLogic;
	}
	@Autowired
	public void setSessionLogic(ICureSessionLogic sessionLogic) {
		this.sessionLogic = sessionLogic;
	}

	@Override
	public Set<String> resolve(org.taktik.icure.dto.filter.contact.ContactByHcPartyTagCodeDateFilter  filter, Filters context) {
		try {
            String hcPartyId = filter.getHealthcarePartyId() != null ? filter.getHealthcarePartyId() : getLoggedHealthCarePartyId();
            HashSet<String> ids = null;

            if (filter.getTagType() != null && filter.getTagCode() != null) {
                ids = new HashSet<>(contactLogic.listServiceIdsByTag(
                        hcPartyId,
                        null, filter.getTagType(),
                        filter.getTagCode(),
						filter.getStartServiceValueDate(), filter.getEndServiceValueDate()));
            }

            if (filter.getCodeType() != null && filter.getCodeCode() != null) {
                List<String> byCode = contactLogic.listServiceIdsByCode(
                        hcPartyId,
                        null, filter.getCodeType(),
                        filter.getCodeCode(),
						filter.getStartServiceValueDate(), filter.getEndServiceValueDate());
                if (ids==null) { ids = new HashSet<>(byCode); } else { ids.retainAll(byCode); }
            }

            if (ids == null) {
                return new HashSet<>(contactLogic.listContactIds(hcPartyId));
            }

            return ids != null ? new HashSet<>(contactLogic.findByServices(ids)) : new HashSet<>();
		} catch (LoginException e) {
			throw new IllegalArgumentException(e);
		}
	}

	private String getLoggedHealthCarePartyId() throws LoginException {
		User user = sessionLogic.getCurrentSessionContext().getUser();
		if (user == null || user.getHealthcarePartyId() == null) {
			throw new LoginException("You must be logged to perform this action. ");
		}
		return user.getHealthcarePartyId();
	}
}

File: src/main/kotlin/org/taktik/icure/asynclogic/impl/filter/contact/ContactByHcPartyTagCodeDateFilter.kt
/*
 *  iCure Data Stack. Copyright (c) 2020 Taktik SA
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful, but
 *     WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *     Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public
 *     License along with this program.  If not, see
 *     <https://www.gnu.org/licenses/>.
 */
package org.taktik.icure.asynclogic.impl.filter.contact

import javax.security.auth.login.LoginException
import kotlinx.coroutines.flow.asFlow
import kotlinx.coroutines.flow.emitAll
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.toList
import org.springframework.stereotype.Service
import org.taktik.icure.asynclogic.AsyncSessionLogic
import org.taktik.icure.asynclogic.ContactLogic
import org.taktik.icure.asynclogic.impl.filter.Filter
import org.taktik.icure.asynclogic.impl.filter.Filters
import org.taktik.icure.domain.filter.contact.ContactByHcPartyTagCodeDateFilter
import org.taktik.icure.entities.Contact
import org.taktik.icure.utils.getLoggedHealthCarePartyId

@Service
class ContactByHcPartyTagCodeDateFilter(private val contactLogic: ContactLogic,
                                        private val sessionLogic: AsyncSessionLogic) : Filter<String, Contact, ContactByHcPartyTagCodeDateFilter> {

    override fun resolve(filter: ContactByHcPartyTagCodeDateFilter, context: Filters) = flow {
        try {
            val hcPartyId: String = filter.healthcarePartyId ?: getLoggedHealthCarePartyId(sessionLogic)
            var ids: HashSet<String>? = null
            if (filter.tagType != null && filter.tagCode != null) {
                ids = HashSet(contactLogic.listContactIdsByTag(
                        hcPartyId,
                        filter.tagType!!,
                        filter.tagCode!!,
                        filter.startOfContactOpeningDate, filter.endOfContactOpeningDate).toList())
            }
            if (filter.codeType != null && filter.codeCode != null) {
                val byCode = contactLogic.listContactIdsByCode(
                        hcPartyId,
                        filter.codeType!!,
                        filter.codeCode!!,
                        filter.startOfContactOpeningDate, filter.endOfContactOpeningDate).toList()
                if (ids == null) {
                    ids = HashSet(byCode)
                } else {
                    ids.retainAll(byCode)
                }
            }
            emitAll(ids?.asFlow() ?: contactLogic.listContactIds(hcPartyId))
        } catch (e: LoginException) {
            throw IllegalArgumentException(e)
        }
    }
}
